flink-table/flink-table-common/src/main/java/org/apache/flink/table/connector/sink/abilities/SupportsPartitioning.java

```diff
@@ -0,0 +1,109 @@
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.flink.table.connector.sink.abilities;
+
+import org.apache.flink.annotation.PublicEvolving;
+import org.apache.flink.table.connector.sink.DynamicTableSink;
+
+import java.util.Map;
+
+/**
+ * Enables to write partitioned data in a {@link DynamicTableSink}.
+ *
+ * <p>Partitions split the data stored in an external system into smaller portions that are identified
+ * by one or more string-based partition keys. A single partition is represented as a {@code Map < String, String >}
+ * which maps each partition key to a partition value. Partition keys and their order are defined by the
+ * catalog table.
+ *
+ * <p>For example, data can be partitioned by region and within a region partitioned by month. The order
+ * of the partition keys (in the example: first by region then by month) is defined by the catalog table. A
+ * list of partitions could be:
+ * <pre>
+ *   List(
+ *     ['region'='europe', 'month'='2020-01'],
+ *     ['region'='europe', 'month'='2020-02'],
+ *     ['region'='asia', 'month'='2020-01'],
+ *     ['region'='asia', 'month'='2020-02']
+ *   )
+ * </pre>
+ *
+ * <p>Given the following partitioned table:
+ * <pre>{@code
+ *   CREATE TABLE t (a INT, b STRING, c DOUBLE, region STRING, month STRING) PARTITION BY (region, month);
+ * }</pre>
+ *
+ * <p>We can insert data into <i>static table partitions</i> using the {@code INSERT INTO ... PARTITION} syntax:
+ * <pre>{@code
+ *     INSERT INTO t PARTITION (region='europe', month='2020-01') SELECT a, b, c FROM my_view;
+ * }</pre>
+ *
+ * <p>If all partition keys get a value assigned in the {@code PARTITION} clause, the operation is considered
+ * as an "insertion into a static partition". In the above example, the query result should be written
+ * into the static partition {@code region='europe', month='2020-01'} which will be passed by the planner
+ * into {@link #applyStaticPartition(Map)}.
+ *
+ * <p>Alternatively, we can insert data into <i>dynamic table partitions</i> using the SQL syntax:
+ * <pre>{@code
+ *     INSERT INTO t PARTITION (region='europe') SELECT a, b, c, month FROM another_view;
+ * }</pre>
+ *
+ * <p>If only a subset of all partition keys (a prefix part) get a value assigned in the {@code PARTITION}
+ * clause, the operation is considered as an "insertion into a dynamic partition". In the above example,
+ * the static partition part is {@code region='europe'} which will be passed by the planner into
+ * {@link #applyStaticPartition(Map)}. The remaining values for partition keys should be obtained from
```

Comment:
nit: Not necessarily related to this PR. Do we have any benefit of having the static partitioning? What is the difference between:
```
 INSERT INTO t PARTITION (region='europe', month="may") SELECT a, b, c FROM another_view;
```
vs
```
 INSERT INTO t SELECT a, b, c, 'europe' as region, `may` as month FROM another_view;
```

Reply:
I think a user can still add records with `asia` value in a `europe` partition.

Reply:
Hmm is that behavior documented somewhere? @lirui-apache @JingsongLi 
As far as I know in oracle it is not possible. There will be a validation step. In hive that's true that there will be no validation step but as far as I know this clause makes sense only for bulk loading when you don't perform any actions on the data, but naively move the whole files (correct me if I am wrong).

This whole discussion should definitely not block the PR. It's just that I find the concepts of partitions a bit unclear and the javadocs here do not help me much.

(Sorry I could not quickly find a good oracle documentation)
https://dev.mysql.com/doc/refman/5.6/en/partitioning-selection.html

>  For statements that insert rows, the behavior differs in that failure to find a suitable partition causes the statement to fail. This is true for both INSERT and REPLACE statements, as shown here:

Reply:
I can also mark this interface `@Experimental` again. I just thought that the interface has stabilized in the the last months. Looking forward to the opinion of @JingsongLi and @lirui-apache

Reply:
I think both syntax suggested by @dawidwys work. But IMHO having an explicit static partition spec makes the semantics a little clearer. For example, by reading the DML, one knows the target table is partitioned and the DML intends to write to some certain partitions. I guess it also makes things easier for the planner because planner needs to pass the static partitions to the sink, and if we don't have the partition spec, it has to iterate all the projection columns to decide which literals are for static partitions.

I don't think hive allows records with `asia` value in a `europe` partition. In hive each partition is a folder and the partition value is encoded in the folder's name. Therefore all data under a `europe` folder belongs to the `europe` partition.

Reply:
As @lirui-apache said, static partitions help sink implementation choose a more efficient `PartitionWriter`.

I am implementing the file system related source and sink. The partition fields is saved in the directory names, and the non-partition field is saved in the data.
So the format writing implementation only sees non-partition fields.